### ISSUE TITLE
Rendering issues on the latest Nightly Build

### ISSUE THREAD

**Issue by Kenny-Hui** [leezer3/OpenBVE#693]
### Description

I am not sure if the latest Nightly Build is meant to be functional, but it seems that there are some severe graphical glitches on the latest Nightly Build. This applies to RouteViewer and In-Game (I haven't tested what happens on ObjectViewer yet) on GL 3.0

NOTE: If you switch to GL 1.2, it only shows this screen:
![image](https://user-images.githubusercontent.com/28094366/131240423-15fec682-7479-4994-93d3-b9cafa40a8ce.png)
(The somewhat corrupted "1" shown is the "Total Alpha Face number", and the number seems to only refresh If I switch from GL3 to GL1)

### Reproduction
Open the route below on RouteViewer or In-Game
Fly Around, and you should see the same graphical glitches in this video:
https://www.youtube.com/watch?v=yzzc1CUEXYE

## Route
A very very simple, incomplete route that can reproduce the bug
[K14.zip](https://github.com/leezer3/OpenBVE/files/7072006/K14.zip)

### Related information 

* Operating system: Windows 10 21H1
* Method of control: Keyboard

**Comment by leezer3** [comment]
Odd.
Don't think I've touched anything there at all recently. Will try and take a look later.

**Comment by leezer3** [comment]
Really odd.
That seems to mainifest if starting in GL3 and switching to GL1.2.

You can clear the issue by pressing F5, investigating at the minute.

**Comment by leezer3** [comment]
Ah.....
https://github.com/leezer3/OpenBVE/pull/595

This switched us to a fully GL3 renderer (previously the text was drawn using GL2 even when in GL3 mode), and this Gl2 text led to the state being coincidentally right for the start of the next frame.

Needed to specififcally activate / deactivate the shader and update the lighting to ensure this when switching renderer types.

**Comment by Kenny-Hui** [comment]
This fixes the rendering switches, but not the "duplicated" face that follows the camera around, it's still the same shown in the yt video linked above.

Now that I am able to properly see stuff in GL 1.2, it seems that the duplicated faces only happens on GL 3.0.
I am not sure what caused this, but maybe it has something to do with object over 25m block length (As I can't find this bug on BVE Route that follows the block length)

**Comment by leezer3** [comment]
Hit a little harder...
(It's actually because your route has so few objects that this is manifesting)

TLDR: It's this little bit of caching:
https://github.com/leezer3/OpenBVE/commit/304df799979a6973e9554ae12ae3535ac15cb713

If we cache the object state matrix last drawn, this speeds up repeated draw calls (e.g. forests, transparencies etc.).
However, in a viewer, camera movement and a single combined object state can cause this cache to be wrong.
